File: middleware/windowmanager/windowmanager.go
package windowmanager

import (
	"context"
	"errors"
	"sync"
	"time"
)

var ErrEmptyMessages = errors.New("messages cannot be empty")
var ErrInvalidReservedTokens = errors.New("reserved_tokens must be less than max_tokens")

type Message struct {
	Role    string
	Content string
	Name    string
}

type ModelSpec struct {
	Name            string
	MaxTokens       int
	ReservedTokens  int
	TokenMultiplier float64
}

type WarningLevel string

const (
	WarningLevelNone     WarningLevel = "none"
	WarningLevelLow      WarningLevel = "low"
	WarningLevelMedium   WarningLevel = "medium"
	WarningLevelHigh     WarningLevel = "high"
	WarningLevelCritical WarningLevel = "critical"
)

type ContextStatus struct {
	UsedTokens      int
	RemainingTokens int
	MaxTokens       int
	ReservedTokens  int
	WarningLevel    WarningLevel
	MessageCount    int
}

type CompactionStrategy interface {
	Compact(messages []Message, targetTokens int, counter TokenCounter) ([]Message, error)
	Name() string
}

type TokenCounter interface {
	Count(messages []Message) (int, error)
	CountMessage(msg Message) (int, error)
}

type ContextWindowManager struct {
	model     ModelSpec
	strategy  CompactionStrategy
	counter   TokenCounter
	mu        sync.RWMutex
	lastCheck time.Time
}

func NewContextWindowManager(model ModelSpec, strategy CompactionStrategy, counter TokenCounter) (*ContextWindowManager, error) {
	if model.MaxTokens <= 0 {
		model.MaxTokens = 4096
	}
	if model.ReservedTokens < 0 {
		model.ReservedTokens = 0
	}
	if model.ReservedTokens >= model.MaxTokens {
		return nil, ErrInvalidReservedTokens
	}
	if model.TokenMultiplier == 0 {
		model.TokenMultiplier = 4.0
	}

	if counter == nil {
		counter = NewDefaultCounter()
	}

	return &ContextWindowManager{
		model:    model,
		strategy: strategy,
		counter:  counter,
	}, nil
}

func (m *ContextWindowManager) Check(ctx context.Context, messages []Message) (*ContextStatus, error) {
	if len(messages) == 0 {
		return &ContextStatus{
			UsedTokens:      0,
			RemainingTokens: m.model.MaxTokens - m.model.ReservedTokens,
			MaxTokens:       m.model.MaxTokens,
			ReservedTokens:  m.model.ReservedTokens,
			WarningLevel:    WarningLevelNone,
			MessageCount:    0,
		}, nil
	}

	m.mu.Lock()
	defer m.mu.Unlock()

	tokenCount, err := m.counter.Count(messages)
	if err != nil {
		return nil, err
	}

	available := m.model.MaxTokens - m.model.ReservedTokens
	remaining := available - tokenCount

	return &ContextStatus{
		UsedTokens:      tokenCount,
		RemainingTokens: remaining,
		MaxTokens:       m.model.MaxTokens,
		ReservedTokens:  m.model.ReservedTokens,
		WarningLevel:    m.calculateWarningLevel(remaining, available),
		MessageCount:    len(messages),
	}, nil
}

func (m *ContextWindowManager) ShouldCompact(ctx context.Context, messages []Message) (bool, error) {
	status, err := m.Check(ctx, messages)
	if err != nil {
		return false, err
	}

	threshold := m.model.MaxTokens / 4
	return status.RemainingTokens < threshold, nil
}

func (m *ContextWindowManager) Compact(ctx context.Context, messages []Message) ([]Message, error) {
	if len(messages) == 0 {
		return messages, nil
	}

	m.mu.Lock()
	defer m.mu.Unlock()

	available := m.model.MaxTokens - m.model.ReservedTokens
	targetTokens := int(float64(available) * 0.75)

	return m.strategy.Compact(messages, targetTokens, m.counter)
}

func (m *ContextWindowManager) CompactIfNeeded(ctx context.Context, messages []Message) ([]Message, bool, error) {
	shouldCompact, err := m.ShouldCompact(ctx, messages)
	if err != nil {
		return nil, false, err
	}

	if !shouldCompact {
		return messages, false, nil
	}

	compacted, err := m.Compact(ctx, messages)
	if err != nil {
		return nil, false, err
	}

	return compacted, true, nil
}

func (m *ContextWindowManager) calculateWarningLevel(remaining, available int) WarningLevel {
	if available == 0 {
		return WarningLevelCritical
	}

	ratio := float64(remaining) / float64(available)

	switch {
	case ratio <= 0.1:
		return WarningLevelCritical
	case ratio <= 0.25:
		return WarningLevelHigh
	case ratio <= 0.5:
		return WarningLevelMedium
	case ratio <= 0.75:
		return WarningLevelLow
	default:
		return WarningLevelNone
	}
}

type LastNCompaction struct {
	KeepCount int
}

func (c *LastNCompaction) Name() string {
	return "last_n"
}

func (c *LastNCompaction) Compact(messages []Message, targetTokens int, counter TokenCounter) ([]Message, error) {
	if len(messages) == 0 {
		return messages, nil
	}

	keepCount := c.KeepCount
	if keepCount <= 0 {
		keepCount = len(messages) / 2
	}
	if keepCount < 1 {
		keepCount = 1
	}

	if keepCount >= len(messages) {
		count, _ := counter.Count(messages)
		if count <= targetTokens {
			return messages, nil
		}
		keepCount = len(messages) / 2
	}

	for i := keepCount; i >= 1; i-- {
		subset := messages[len(messages)-i:]
		count, err := counter.Count(subset)
		if err != nil {
			continue
		}
		if count <= targetTokens {
			return subset, nil
		}
	}

	return messages[len(messages)-1:], nil
}

type SummaryCompaction struct {
	summaryPrompt string
	maxSummaryLen int
}

func NewSummaryCompaction() *SummaryCompaction {
	return &SummaryCompaction{
		summaryPrompt: "Summarize the following conversation concisely, preserving key information:",
		maxSummaryLen: 500,
	}
}

func (c *SummaryCompaction) Name() string {
	return "summary"
}

func (c *SummaryCompaction) Compact(messages []Message, targetTokens int, counter TokenCounter) ([]Message, error) {
	if len(messages) == 0 {
		return messages, nil
	}

	var systemMsg *Message
	var otherMsgs []Message

	for i, msg := range messages {
		if msg.Role == "system" {
			systemMsg = &messages[i]
		} else {
			otherMsgs = append(otherMsgs, msg)
		}
	}

	if len(otherMsgs) == 0 {
		return messages, nil
	}

	count, _ := counter.Count(otherMsgs)
	if count <= targetTokens {
		if systemMsg != nil {
			result := append([]Message{*systemMsg}, otherMsgs...)
			return result, nil
		}
		return messages, nil
	}

	var result []Message
	if systemMsg != nil {
		result = append(result, *systemMsg)
	}

	summaryTokens := targetTokens / 4
	summaryChars := int(float64(summaryTokens) * 3)

	summary := c.summarizeMessages(otherMsgs, summaryChars)
	summaryMsg := Message{
		Role:    "system",
		Content: "[Previous conversation summarized: " + summary + "]",
	}
	result = append(result, summaryMsg)

	currentCount, _ := counter.Count(result)
	if currentCount > targetTokens {
		keepCount := len(otherMsgs) / 3
		if keepCount < 1 {
			keepCount = 1
		}
		recentMsgs := otherMsgs[len(otherMsgs)-keepCount:]
		result = append(result, recentMsgs...)
	}

	return result, nil
}

func (c *SummaryCompaction) summarizeMessages(messages []Message, maxLen int) string {
	var content string
	for _, msg := range messages {
		prefix := msg.Role
		if msg.Name != "" {
			prefix = msg.Name
		}
		content += prefix + ": " + msg.Content + "\n"
	}

	if len(content) <= maxLen {
		return content
	}

	truncated := content[:maxLen-3]
	lastSpace := -1
	for i := len(truncated) - 1; i >= 0; i-- {
		if truncated[i] == ' ' || truncated[i] == '\n' {
			lastSpace = i
			break
		}
	}

	if lastSpace > 0 {
		return truncated[:lastSpace] + "..."
	}

	return truncated + "..."
}

type Priority int

const (
	PrioritySystem Priority = iota
	PriorityHigh
	PriorityMedium
	PriorityLow
)

type PriorityMessage struct {
	Message
	Priority Priority
}

type PriorityBasedCompaction struct {
	keepSystem bool
}

func NewPriorityBasedCompaction() *PriorityBasedCompaction {
	return &PriorityBasedCompaction{
		keepSystem: true,
	}
}

func (c *PriorityBasedCompaction) Name() string {
	return "priority"
}

func (c *PriorityBasedCompaction) Compact(messages []Message, targetTokens int, counter TokenCounter) ([]Message, error) {
	if len(messages) == 0 {
		return messages, nil
	}

	var priorityMsgs []PriorityMessage
	for _, msg := range messages {
		priority := PriorityMedium
		switch msg.Role {
		case "system":
			if c.keepSystem {
				priority = PrioritySystem
			} else {
				priority = PriorityLow
			}
		case "assistant":
			priority = PriorityHigh
		case "user":
			priority = PriorityMedium
		default:
			priority = PriorityLow
		}
		priorityMsgs = append(priorityMsgs, PriorityMessage{Message: msg, Priority: priority})
	}

	sorted := make([]PriorityMessage, len(priorityMsgs))
	copy(sorted, priorityMsgs)

	for i := 0; i < len(sorted)-1; i++ {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[i].Priority > sorted[j].Priority {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}

	var result []Message
	for _, pm := range sorted {
		result = append(result, pm.Message)
		count, _ := counter.Count(result)
		if count > targetTokens {
			result = result[:len(result)-1]
			break
		}
	}

	if len(result) == 0 && len(messages) > 0 {
		result = append(result, messages[len(messages)-1])
	}

	return result, nil
}

type DefaultCounter struct{}

func NewDefaultCounter() *DefaultCounter {
	return &DefaultCounter{}
}

func (c *DefaultCounter) Count(messages []Message) (int, error) {
	total := 0
	for _, msg := range messages {
		count, err := c.CountMessage(msg)
		if err != nil {
			return 0, err
		}
		total += count
	}
	return total, nil
}

func (c *DefaultCounter) CountMessage(msg Message) (int, error) {
	content := msg.Role
	if msg.Name != "" {
		content += msg.Name
	}
	content += msg.Content

	tokens := len(content) / 4
	return tokens + 4, nil
}

File: middleware_py/windowmanager.py
"""Context window management for preemptive context monitoring."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Protocol


class WarningLevel(str, Enum):
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


@dataclass
class Message:
    role: str
    content: str
    name: str = ""


@dataclass
class ModelSpec:
    name: str = ""
    max_tokens: int = 4096
    reserved_tokens: int = 0
    token_multiplier: float = 4.0


@dataclass
class ContextStatus:
    used_tokens: int
    remaining_tokens: int
    max_tokens: int
    reserved_tokens: int
    warning_level: WarningLevel
    message_count: int


class CompactionStrategy(Protocol):
    """Protocol for compaction strategies."""

    @abstractmethod
    def compact(self, messages: list[Message], target_tokens: int, counter: "TokenCounter") -> list[Message]:
        """Compact messages to fit within target token limit."""
        ...

    @abstractmethod
    def name(self) -> str:
        """Return strategy name."""
        ...


class TokenCounter(Protocol):
    """Protocol for token counting."""

    @abstractmethod
    def count(self, messages: list[Message]) -> int:
        """Count total tokens in messages."""
        ...

    @abstractmethod
    def count_message(self, message: Message) -> int:
        """Count tokens in a single message."""
        ...


class ContextWindowManager:
    """Manages context window with preemptive compaction."""

    def __init__(
        self,
        model: ModelSpec,
        strategy: CompactionStrategy | None = None,
        counter: TokenCounter | None = None,
    ):
        if model.max_tokens <= 0:
            model.max_tokens = 4096
        if model.reserved_tokens < 0:
            model.reserved_tokens = 0
        if model.reserved_tokens >= model.max_tokens:
            raise ValueError("reserved_tokens must be less than max_tokens")
        if model.token_multiplier == 0:
            model.token_multiplier = 4.0

        self._model = model
        self._strategy = strategy or LastNCompaction()
        self._counter = counter or DefaultCounter()
        self._last_check: datetime | None = None

    def check(self, messages: list[Message]) -> ContextStatus:
        """Check current context status."""
        if not messages:
            return ContextStatus(
                used_tokens=0,
                remaining_tokens=self._model.max_tokens - self._model.reserved_tokens,
                max_tokens=self._model.max_tokens,
                reserved_tokens=self._model.reserved_tokens,
                warning_level=WarningLevel.NONE,
                message_count=0,
            )

        token_count = self._counter.count(messages)
        available = self._model.max_tokens - self._model.reserved_tokens
        remaining = available - token_count

        return ContextStatus(
            used_tokens=token_count,
            remaining_tokens=remaining,
            max_tokens=self._model.max_tokens,
            reserved_tokens=self._model.reserved_tokens,
            warning_level=self._calculate_warning_level(remaining, available),
            message_count=len(messages),
        )

    def should_compact(self, messages: list[Message]) -> bool:
        """Check if compaction is needed."""
        status = self.check(messages)
        threshold = self._model.max_tokens // 4
        return status.remaining_tokens < threshold

    def compact(self, messages: list[Message]) -> list[Message]:
        """Compact messages to free up space."""
        if not messages:
            return messages

        available = self._model.max_tokens - self._model.reserved_tokens
        target_tokens = int(available * 0.75)

        return self._strategy.compact(messages, target_tokens, self._counter)

    def compact_if_needed(self, messages: list[Message]) -> tuple[list[Message], bool]:
        """Compact messages if needed, returns (messages, did_compact)."""
        if not self.should_compact(messages):
            return messages, False

        compacted = self.compact(messages)
        return compacted, True

    def _calculate_warning_level(self, remaining: int, available: int) -> WarningLevel:
        """Calculate warning level based on remaining tokens."""
        if available == 0:
            return WarningLevel.CRITICAL

        ratio = remaining / available

        if ratio <= 0.1:
            return WarningLevel.CRITICAL
        elif ratio <= 0.25:
            return WarningLevel.HIGH
        elif ratio <= 0.5:
            return WarningLevel.MEDIUM
        elif ratio <= 0.75:
            return WarningLevel.LOW
        else:
            return WarningLevel.NONE


class LastNCompaction:
    """Compaction strategy that keeps the last N messages."""

    def __init__(self, keep_count: int = 0):
        self._keep_count = keep_count

    def name(self) -> str:
        return "last_n"

    def compact(self, messages: list[Message], target_tokens: int, counter: TokenCounter) -> list[Message]:
        if not messages:
            return messages

        keep_count = self._keep_count
        if keep_count <= 0:
            keep_count = len(messages) // 2
        if keep_count < 1:
            keep_count = 1

        if keep_count >= len(messages):
            count = counter.count(messages)
            if count <= target_tokens:
                return messages
            keep_count = len(messages) // 2

        for i in range(keep_count, 0, -1):
            subset = messages[len(messages) - i:]
            count = counter.count(subset)
            if count <= target_tokens:
                return subset

        return [messages[-1]]


class Priority(str, Enum):
    SYSTEM = "system"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class PriorityMessage:
    message: Message
    priority: Priority


class PriorityBasedCompaction:
    """Compaction strategy that keeps messages by priority."""

    def __init__(self, keep_system: bool = True):
        self._keep_system = keep_system

    def name(self) -> str:
        return "priority"

    def compact(self, messages: list[Message], target_tokens: int, counter: TokenCounter) -> list[Message]:
        if not messages:
            return messages

        priority_msgs: list[PriorityMessage] = []
        for msg in messages:
            if msg.role == "system":
                priority = Priority.SYSTEM if self._keep_system else Priority.LOW
            elif msg.role == "assistant":
                priority = Priority.HIGH
            elif msg.role == "user":
                priority = Priority.MEDIUM
            else:
                priority = Priority.LOW
            priority_msgs.append(PriorityMessage(message=msg, priority=priority))

        priority_order = [Priority.SYSTEM, Priority.HIGH, Priority.MEDIUM, Priority.LOW]
        sorted_msgs = sorted(priority_msgs, key=lambda pm: priority_order.index(pm.priority))

        result: list[Message] = []
        for pm in sorted_msgs:
            result.append(pm.message)
            if counter.count(result) > target_tokens:
                result = result[:-1]
                break

        if not result and messages:
            result = [messages[-1]]

        return result


class SummaryCompaction:
    """Compaction strategy that summarizes older messages."""

    def __init__(self, max_summary_len: int = 500):
        self._max_summary_len = max_summary_len

    def name(self) -> str:
        return "summary"

    def compact(self, messages: list[Message], target_tokens: int, counter: TokenCounter) -> list[Message]:
        if not messages:
            return messages

        system_msg: Message | None = None
        other_msgs: list[Message] = []

        for msg in messages:
            if msg.role == "system":
                system_msg = msg
            else:
                other_msgs.append(msg)

        if not other_msgs:
            return messages

        count = counter.count(other_msgs)
        if count <= target_tokens:
            if system_msg:
                return [system_msg] + other_msgs
            return messages

        result: list[Message] = []
        if system_msg:
            result.append(system_msg)

        summary_tokens = target_tokens // 4
        summary_chars = int(summary_tokens * 3)

        summary = self._summarize_messages(other_msgs, summary_chars)
        summary_msg = Message(
            role="system",
            content=f"[Previous conversation summarized: {summary}]",
        )
        result.append(summary_msg)

        if counter.count(result) > target_tokens:
            keep_count = len(other_msgs) // 3
            if keep_count < 1:
                keep_count = 1
            result.extend(other_msgs[-keep_count:])

        return result

    def _summarize_messages(self, messages: list[Message], max_len: int) -> str:
        """Create a simple summary of messages."""
        content = ""
        for msg in messages:
            prefix = msg.name if msg.name else msg.role
            content += f"{prefix}: {msg.content}\n"

        if len(content) <= max_len:
            return content

        truncated = content[: max_len - 3]
        last_space = truncated.rfind(" ")
        if last_space > 0:
            return truncated[:last_space] + "..."

        return truncated + "..."


class DefaultCounter:
    """Default token counter using character-based estimation."""

    def count(self, messages: list[Message]) -> int:
        total = 0
        for msg in messages:
            total += self.count_message(msg)
        return total

    def count_message(self, message: Message) -> int:
        content = message.role
        if message.name:
            content += message.name
        content += message.content

        tokens = len(content) // 4
        return tokens + 4